### PR TITLE
added max file limit

### PR DESCRIPTION
**Pull request by heejaechang**
**Customer scenario**

User has big file (over 100MB) in solution and VS crashes OOM.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/16796

**Workarounds, if any**

remove the massive file from solution or split the file into smaller chunk

**Risk**

since we are now limiting size of text file (100MB) we will read, if someone had a big file which is big but not big enought to make VS OOM, that file will be no longer read into VS (Roslyn).

but we provide regkey to change default threshold, so we do still provide a workaround for such user.

**Performance impact**

there should be no change on performance

**Is this a regression from a previous update?**

No

**Root cause analysis:**

we need to read in whole file into memory to operate, we do chunk them so that it doesn't go to the LOH, but still we read whole thing in memory to operate rather than reading in or map portion of a file to operate. but naturally that sets limit on how much big file we can support, especially since devenv is in 32bit process. this fix explicitly set limit so that we reduce chance of going OOM.

since we at least chunk, we won't OOM due to lack of giant continuous memory block.

**How was the bug found?**

customer report.

### CODE REVIEW

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 1]
```diff
@@ -372,6 +372,21 @@ internal static DateTime GetFileTimeStamp(string fullPath)
...
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+            try
+            {
+                var info = new FileInfo(fullPath);
+                return info.Length;
+            }
+            catch (Exception e)
+            {
+                throw new IOException(e.Message, e);
```
Comment (Pilchie) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Do you want to just use http://source.roslyn.io/Microsoft.CodeAnalysis/R/9c1e0a07e83f2ac3.html instead?

(Also, should that be changed to use an exception filter to avoid ever catching IOException in the first place)?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 2]
```diff
@@ -7,11 +7,20 @@
...
+using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    internal static class FileTextLoaderOptions
+    {
+        [ExportOption]
+        internal static readonly Option<long> FileLengthThreshold = new Option<long>(nameof(FileTextLoaderOptions), nameof(FileLengthThreshold), defaultValue: 10 * 1024 * 1024,
```
Comment (Pilchie) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Per mail-thread, I don't think this is large enough.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 3]
```diff
@@ -198,5 +211,18 @@ private string GetDebuggerDisplay()
...
             return nameof(Path) + " = " + Path;
         }
+
+        private static void ValidateFileLength(Workspace workspace, string path)
+        {
+            var fileLength = FileUtilities.GetFileLength(path);
+            if (fileLength > workspace.Options.GetOption(FileTextLoaderOptions.FileLengthThreshold))
+            {
+                // log max file length to AI
+                Logger.Log(FunctionId.FileTextLoader_FileLengthThresholdExceeded, KeyValueLogMessage.Create(m => m["FileLength"] = fileLength));
```
Comment (Pilchie) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Since this path will also be hit by `AdditionalDocuments`, is there something more actionable than just size that we can log here without capturing PII?  File extension at least?

File: src/Workspaces/Core/Portable/WorkspacesResources.resx [hunk 4]
```diff
@@ -610,4 +610,7 @@
   <data name="The_first_word_0_must_begin_with_a_lower_case_character" xml:space="preserve">
     <value>The first word, '{0}', must begin with a lower case character</value>
   </data>
+  <data name="File_size_exceeds_maximum_allowed_size_0" xml:space="preserve">
+    <value>File size exceeds maximum allowed size: {0}</value>
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/WorkspacesResources.resx]: We're past loc cutoff, so it might be better to just use a hardcoded message in RTW, and update later.  Not sure though.

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 1 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: ah. ya. will change it to use that.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 2 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: sure

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 3 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: sure we can do that.

File: src/Workspaces/Core/Portable/WorkspacesResources.resx (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/WorkspacesResources.resx]: okay

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Why do we wrap in this way?  My intuition is "just throw because catching non-Exception exceptions is trying to win a game of whack a mole".

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 5]
```diff
@@ -198,5 +211,18 @@ private string GetDebuggerDisplay()
...
+
+                var message = string.Format(WorkspacesResources.File_size_exceeds_maximum_allowed_size_0, path);
+                throw new IOException(message);
```
Comment (jaredpar) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Why IOExecption?  Why not just throw Execption.  Or InvalidOperationException.  No IO is actually occuring here.

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: We consistently distinguish between IOExceptions and anything else across the code base.

Comment (jaredpar) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: This change would argue that distinction is very questionable.

Comment (tmat) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: IOException means external error as opposed to our bug.

Comment (jaredpar) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Sure.  It's only useful though if we are diligent and wrap every single place which throws an exception with IOException.  That's a whack a mole problem.  Every time we find a new place, wrap wrap warp.

Much simpler: eliminate the middle man and just catch exception.  The end result is indistinguishable.

Comment (tmat) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Yes, we are diligent and are wrapping them. It's worked well so far.

Catch where?

Comment (jaredpar) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Sure.  We are diligent to the places we know about and that's the problem.  There are plenty of more cases where IO occurs that we don't know about.  Every miss is just a crash that we're passing onto the customer instead of fixing now.  

This is why I'm always very skeptical of approaches like this.  Every time a customer reports the exception we will just catch ,wrap in IOException and move on.  Hence the end behavior is indistinguishable from just catching Exception in the first place.  This approach just causes us more work.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 5 above)
Comment (tmat) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: +1 I don't think we should be throwing an exception here. I'd either change it to Try-pattern, or define a new exception type  that's gonna be handled here: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/TextDocumentState.cs,177

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Yes, we need to be careful whenever we call outside of our code base. I'm not saying it's easy. But the solution is not to just catch and ignore all exceptions everywhere. That's a nightmare - things will stop working and nobody will know why. The customer won't be able to create a small repro and we won't be able to fix anything. So yes, you won't have VS crashing. But you won't have it working either. Better approach is to catch all exceptions in a filter with NFW reported on some recovery boundaries where we can report an error and continue running with no (best effort) state corruption.

Comment (heejaechang) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: @jaredpar @tmat so what is the conclusion? I am thinking not changing FileUtilities and take file length thing in to FileTextLoader itself.

Comment (tmat) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: I'd leave it here and do what we do in the rest of the FileUtilities - which is to turn non-IOExceptions to IOExceptions. It's consistent with the rest of the utils code.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 5 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: @Pilchie how do you think. looks like FileTextLoader consider any exception happen inside as IOException regardless they are generally considered IOException or not in OS standpoint.

**Comment by heejaechang** [comment]
ping?

**Comment by heejaechang** [comment]
@jaredpar @tmat @dotnet/roslyn-compiler @AlekseyTs @cston @VSadov can I get code review for changes in FileUtilities?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 6]
```diff
@@ -7,11 +7,21 @@
...
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    internal static class FileTextLoaderOptions
+    {
+        // File Threshold 100MB
+        [ExportOption]
+        internal static readonly Option<long> FileLengthThreshold = new Option<long>(nameof(FileTextLoaderOptions), nameof(FileLengthThreshold), defaultValue: 100 * 1024 * 1024,
+            storageLocations: new LocalUserProfileStorageLocation(@"Roslyn\Internal\Performance\Text\FileLengthThreshold"));
```
Comment (jaredpar) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: The use of "Internal" here seemed wrong given users may need to change it.  Can't tell if this is an established pattern though.  If it is then okay.

Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: user doesn't change it through code, but through regkey, so option itself is internal.

**Comment by jac009** [comment]
@heejaechang  It's not clear to me that this has the telemetry of hitting the issue from my review.  Is that just the "Log" statements?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 6 above)
Comment (jaredpar) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Understood.  I'm referring to the use of Internal in the reg key name.

**Comment by AlekseyTs** [comment]
FileUtilities LGTM.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 6 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: ah, that one. ya, we put all reg key which is not public through tool->options, there.

**Comment by heejaechang** [comment]
@jac009 Yes. Log is the one that send telemetry to AI.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 7]
```diff
@@ -198,5 +212,36 @@ private string GetDebuggerDisplay()
...
+                    m["FileLength"] = fileLength;
+                    m["Ext"] = TryGetFileExtension(path);
+                }));
+
+                var message = string.Format(WorkspacesResources.File_size_exceeds_maximum_allowed_size_0, path);
+                throw new InvalidDataException(message);
+            }
+        }
+
+        private static string TryGetFileExtension(string path)
```
Comment (tmat) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Nit: I believe we already have a utility function to do exactly this: ```PathUtilities.GetExtension```

**Comment by tmat** [comment]
👍

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 7 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: ah, okay.

**Comment by heejaechang** [comment]
@Pilchie should I ask matt for bar check for RTW?

**Comment by heejaechang** [comment]
@natidea just FYI, looks like we will bring this for RTW.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 8]
```diff
@@ -198,5 +212,22 @@ private string GetDebuggerDisplay()
...
+        private static void ValidateFileLength(Workspace workspace, string path)
+        {
+            var fileLength = FileUtilities.GetFileLength(path);
+            if (fileLength > workspace.Options.GetOption(FileTextLoaderOptions.FileLengthThreshold))
+            {
+                // log max file length to AI
+                Logger.Log(FunctionId.FileTextLoader_FileLengthThresholdExceeded, KeyValueLogMessage.Create(m =>
+                {
+                    m["FileLength"] = fileLength;
+                    m["Ext"] = PathUtilities.GetExtension(path);
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Will this get associated with some other project type information? My worry is there's some customer out there with some large auto-generated file (think LINQ, EF, etc.) and we'll see here it's a C# file but not know further details. Although I'm not sure what else _could_ be logged here...

File: src/Workspaces/Core/Portable/WorkspacesResources.resx (hunk 4 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/WorkspacesResources.resx]: Should this say what the file size was? Something like "File '{0}' size of {1} exceeds maximum allowed size of {2}"? Just so that way if a customer does hit that they have some hint what they should try getting their file under, vs. just going "uh, I need to make it smaller I guess?"

**Comment by jasonmalinowski** [comment]
Tagging @dotnet/roslyn-ide since this is touching workspaces.

**Comment by CyrusNajmabadi** [comment]
Yes. Please always tag @dotnet/roslyn-ide if we're making changes that touch infrastructural pieces.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 9]
```diff
@@ -7,11 +7,21 @@
...
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    internal static class FileTextLoaderOptions
+    {
+        // File Threshold 100MB
+        [ExportOption]
+        internal static readonly Option<long> FileLengthThreshold = new Option<long>(nameof(FileTextLoaderOptions), nameof(FileLengthThreshold), defaultValue: 100 * 1024 * 1024,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: I don't know what FileLengthThreshold means.  Can we comment this.  What happens at 100MB?  Do we simply ignore hte file entirely?  Do we have the document, but we truncate it at 100MB?  Do we have the document, but treat its contents as empty?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 10]
```diff
@@ -198,5 +212,22 @@ private string GetDebuggerDisplay()
         {
             return nameof(Path) + " = " + Path;
         }
+
+        private static void ValidateFileLength(Workspace workspace, string path)
+        {
+            var fileLength = FileUtilities.GetFileLength(path);
+            if (fileLength > workspace.Options.GetOption(FileTextLoaderOptions.FileLengthThreshold))
+            {
+                // log max file length to AI
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: What is AI?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 11]
```diff
@@ -198,5 +212,22 @@ private string GetDebuggerDisplay()
...
+                Logger.Log(FunctionId.FileTextLoader_FileLengthThresholdExceeded, KeyValueLogMessage.Create(m =>
+                {
+                    m["FileLength"] = fileLength;
+                    m["Ext"] = PathUtilities.GetExtension(path);
+                }));
+
+                var message = string.Format(WorkspacesResources.File_size_exceeds_maximum_allowed_size_0, path);
+                throw new InvalidDataException(message);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: How will this manifest higher up the stack?  What is the end user experience this causes?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 8 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: @jasonmalinowski  non-fatal watson then?

File: src/Workspaces/Core/Portable/WorkspacesResources.resx (hunk 4 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/WorkspacesResources.resx]: sure. @Pilchie who is localization champ by the way?

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 12]
```diff
@@ -366,6 +366,29 @@ internal static DateTime GetFileTimeStamp(string fullPath)
             {
                 return File.GetLastWriteTimeUtc(fullPath);
             }
+            catch (IOException)
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: I used FileUtilities since that was the one already used in this file. just looked PerformIO which seems very similar to FileUtilites, not sure why there is 2 utilities which does same thing.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 9 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: sure

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 10 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: application insight

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 11 above)
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: same as other failure in this file. showing in workspace failure output pane.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 10 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Please don't abbr unnecessarily.

**Comment by heejaechang** [comment]
@CyrusNajmabadi have you read description?

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs (hunk 11 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: Please comment code to make that clear.  It's also unclear to me what the user does at that point.  Is the workspace unusable?  Do they have any recourse on what they can do?

**Comment by CyrusNajmabadi** [comment]
Yes.  THe description is extremely vague and unclear.

**Comment by heejaechang** [comment]
@CyrusNajmabadi you should ask @Pilchie and @MattGertz why it is appropriate fix since they are the one decide to do this.

**Comment by heejaechang** [comment]
@CyrusNajmabadi hmm.. so want me to add comment on which part? on 100MB limit or what happen it throws?

**Comment by CyrusNajmabadi** [comment]
I'm asking for the description (and code) to actually explain what is going on.  Reading the code, i cannot tell what the actual behavior of tihs will be.  The code is inadequately commented, and i cannot sign off that this is any sort of appropriate fix.

Please:

1) State what behavior we have now.
2) State what behavior we want to have.
3) Explain how this change produces the behavior we want.

**Comment by heejaechang** [comment]
@CyrusNajmabadi in description or in code? where in commit description?

**Comment by MattGertz** [comment]
Actually,  it was JoC 's request...

**Comment by Pilchie** [comment]
Code is better than commit description, since it's right there later.

**Comment by CyrusNajmabadi** [comment]
> in description or in code

Both are fine.  but, at the very least, this needs to be in the code.

> where

Ideally in any codepath that performs these checks.  People reading the code need to understand what the checks are doing and what impact it has on the workspace.  For example (as i asked earlier):

1. Is the workspace usable after this?
2. What impact does this have in the document in the workspace?  i.e. is the document in the workspace?  Is it there with truncated text?  Is it there with empty text?  etc. etc.
3. What options does the user have at this point?  Do they need to close the project?  Do they need to close VS?  etc. etc.

Basically, *nothing* has actually been *explained* here.  I want to know *precisely* what the impact of this change is; what we want to accomplish; and how this PR gets us that.

--

Note: this PR may be perfect (as far as the code change goes).  I'm simply stating that i have no way to effectively review this change as i do not know what you're trying to accomplish, and thus i have no idea if your code change is appropriate.

**Comment by heejaechang** [comment]
@CyrusNajmabadi sure, trying to add comment. just to want to make sure where you want. since it is just throwing, it is other part that decide what to do with the caught exception. so, if I put comment what workspace does because of this exception in the code area I changed, it is far part from where it actually happens. so, if later some body change that behavior, that person will have no idea, the comment exist here which describe workspace behavior he is changing.

basically this code has nothing to do with 1,2,3. it is the one who caught the exception decide 1,2,3 which is in DocumentState I believe.

**Comment by CyrusNajmabadi** [comment]
Sorry, i cannot sign off without a complete understanding of the impact of this change.  It is not sufficient ot say (effectively) "i am throwing, and it is someone else's responsibility as to what happens then".  I need the actual information as to what impact this change will have.  Otherwise, taking this fix would be very irresponsible as it could cause any number of bad things to happen, right when we're in escrow.

As for the comment location.  If you feel like there are two distant piece of code that can be impacted by this, then feel free to comment both locations.  i.e. in this location that you're changing, comment on which parts of the workspace that are affected by this.  And in the workspace, please comment on which parts have to deal with the potential exceptions that can be thrown here.

> that person will have no idea

If there is ever a concern about some 'person' having 'no idea' about what's happening, then the onus is on us to effectively comment hte places of concern so that that doesn't happen.

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 12 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: opened tracking issue - https://github.com/dotnet/roslyn/issues/16836

**Comment by heejaechang** [comment]
@CyrusNajmabadi updated comments!

File: src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs [hunk 13]
```diff
@@ -1417,6 +1418,66 @@ public void TestProjectCompletenessWithMultipleProjects()
...
+            // set max file length to 1 bytes
+            var maxLength = 1;
+            var workspace = new AdhocWorkspace(TestHost.Services, ServiceLayer.Host);
+            workspace.Options = workspace.Options.WithChangedOption(FileTextLoaderOptions.FileLengthThreshold, maxLength);
+
+            string tempPath = null;
+            try
+            {
+                // save temp file
+                tempPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
```
Comment (tmat) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: Or even better: [Temp.CreateDirectory()](http://source.roslyn.io/#Roslyn.Test.Utilities/TempFiles/TempRoot.cs,b69f390853e89250,references)

Comment (tmat) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: And remove try-finally.

Comment (heejaechang) [inline src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs]: sure, will change to use that. I searched how other does this kind of thing in code base and did same thing.

**Comment by heejaechang** [comment]
ping?

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs (hunk 12 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs]: Thanks  #closed.

File: src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs [hunk 14]
```diff
@@ -198,5 +219,30 @@ private string GetDebuggerDisplay()
         {
             return nameof(Path) + " = " + Path;
         }
+
+        private static void ValidateFileLength(Workspace workspace, string path)
+        {
+            // Validate file length is under our threshold. 
+            // Otherwise, rather than reading the content into the memory, we will throw
+            // InvalidDataException to caller of FileTextLoader.LoadText to deal with
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/FileTextLoader.cs]: can we add this exception to the documentation of the virtual calls?

**Comment by CyrusNajmabadi** [comment]
👍 Thanks for adding the better documentation!

**Comment by natidea** [comment]
Approved to merge via [377245](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/377245)
